kvserver: add spanSetWriteBatch to spanset package

iskettaneh opened the pull request:
This PR adds spanSetWriteBatch to the spanset package and asserts on accesses. This will be used later by an engine wrapper to assert on the engine accesses.

References: #158281

Release notes: None

iskettaneh committed:
kvserver: integrate spanSetWriteBatch into SpansetBatch

This commit integrates spanSetWriteBatch into SpansetBatch,
which will allow a future PR to assert all engine accesses
by wrapping Reader, Writer, ReadWriter, Batch with the
spanset type that does the assertions.

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -855,6 +855,106 @@ func (s spanSetBatch) shallowCopy() *spanSetBatch {
+	if err != nil {
+		return err
+	}
+	if err := s.spanSetWriter.checkAllowed(mvccKey.Key); err != nil {
```
IDK if this sometimes should be `checkAllowedAt`, WDYT? I still don't have a good grasp on the "at" variations, and how to reason about their correctness.

iskettaneh replied:
we get this for free. The spansetwriter does that.
It's not very clear to me as well given that we don't really use the timestamps in the assertions, but I followed the same previous pattern so that if in the future we decide to strengthen our assertions, it's gonna be a bit easier.
<!-- Reviewable comment -Ogr4nAe9_r3CTUltCj_:byfk2da -->
<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -855,6 +855,106 @@ func (s spanSetBatch) shallowCopy() *spanSetBatch {
+// PutInternalRangeKey implements storage.InternalWriter.
+func (s *spanSetWriteBatch) PutInternalRangeKey(start, end []byte, key rangekey.Key) error {
+	// Decode the engine keys to check spans.
+	startKey, err := storage.DecodeMVCCKey(start)
```
Should this be `DecodeEngineKey`? IDK if these keys are always MVCC keys / what to expect here.

Same in a few other places.

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -855,6 +855,106 @@ func (s spanSetBatch) shallowCopy() *spanSetBatch {
 
+// spanSetWriteBatch wraps a storage.WriteBatch and adds span checking.
+type spanSetWriteBatch struct {
+	spanSetWriter
```
Should we move the implementation of `InternalWriter` methods into `spanSetWriter`, so that it's written once?

Seems doable if [here](https://github.com/cockroachdb/cockroach/blob/5b180f5c475e42a55f3607f1a994ee9b028a897a/pkg/kv/kvserver/spanset/batch.go#L531) we say `w storage.InternalWriter` .

pav-kv replied:
There is probably some interface/type restructuring that would help. The hierarchical interfaces make it harder. Here is some alternative perspective:

Imagine all the non-overlapping "sub-interfaces" that we need. Then every interface is a sum of these. So if we have K sub-interfaces, we can compose 2^K different ones. We can do the same with the implementation.

Example: `storage.InternalWriter` is wrapped around `storage.Writer`. Let's imagine that `InternalWriter` does not embed `Writer`, so these are non-overlapping. Then let's imagine we have two types:

```golang
type spanSetWriter struct  { } // implements Writer
type spanSetInternalWriter { } // implements InternalWriter (without Writer)
```

Then we can compose stuff:

```golang
type spanSetBatch struct {
	spanSetWriter
	spanSetInternalWriter
	b storage.Batch
}

type spanSetWriteBatch struct {
	spanSetWriter
	spanSetInternalWriter
	b storage.WriteBatch
}
```

Note that we don't even need to declare this new interface, or change ones in `storage`. We just make types with non-overlapping methods and compose them; they magically fit whatever hierarchy of interfaces exists.

pav-kv replied:
There is also an approach with less of a paradigm shift:

```golang
// just need to add 3 methods
type spanSetInternalWriter struct {
	spanSetWriter
}

type spanSetBatch struct {
	spanSetInternalWriter
	b storage.Batch
}

type spanSetWriteBatch struct {
	spanSetInternalWriter
	b storage.WriteBatch
}
```

iskettaneh replied:
I flattened the structs a bit to have more small pieces that can fit together.
<!-- Reviewable comment -Ogr4Zkc7XVNYUNx51e2:bw95zaa -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> @iskettaneh made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159847)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv).



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -855,6 +856,115 @@ func (s spanSetBatch) shallowCopy() *spanSetBatch {
 	return &b
 }
 
+type spanSetInternalWriter struct {
```
It would be a bummer though if this doesn't work:

```golang
type spanSetBatch {
	spanSetReader // inherit everything from Reader
	spanSetWriteBatch // inherit everything from WriteBatch
	b storage.Batch // add Batch-specific stuff not covered
}
```

due to some field name conflicts (both `spanSetReader` and `spanSetWriteBatch` having `spans` etc inside). I'm forgetting how this works in Go.

iskettaneh reviewed (commented):
> @iskettaneh made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159847)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv).



<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
Quick comment before I make another pass: I expected a smaller diff, literally a new type + implementation of most methods moving from `spanSetBatch` to `spanSetWriteBatch`.

Is there a method implemented twice (except `Close`) that needs to be removed? The various ones like `PutInternalRangeKey` seem like candidates - I thought they were already supported in one of your previous PRs. But maybe not. An easier review would be if these were separate commits, but if it's too hard don't do it.

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -897,9 +980,9 @@ func DisableReaderAssertions(reader storage.Reader) storage.Reader {
 	switch v := rw.(type) {
 	case ReadWriter:
-		return DisableReadWriterAssertions(v.w.(storage.ReadWriter))
+		return DisableReadWriterAssertions(v.spanSetWriter.w.(storage.ReadWriter))
```
Hm, does this work? `v.spanSetWriter.w` is a `storage.Writer`, and we're converting it to `storage.ReadWriter`. Is Go somehow able to up-cast this interface?

pav-kv replied:
Interesting, seems to work: https://go.dev/play/p/_5jPANdw17p

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -799,83 +814,151 @@ func (s spanSetBatch) NewBatchOnlyMVCCIterator(
+}
+
+// spanSetWriteBatch wraps a storage.WriteBatch and adds span checking.
+type spanSetWriteBatch struct {
```
nit: put this type above `spanSetBatch`, as this file is ordered from more basic to bigger interfaces. This might have also (or not) made the diff cleaner / easier to review (mostly signature changes).

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -764,17 +765,25 @@ func NewReader(r storage.Reader, spans *SpanSet, ts hlc.Timestamp) storage.Reade
+// NewReadWriterAt returns a storage.ReadWriter that asserts access of the
+// underlying ReadWriter against the given SpanSet.
+//
+// NewReadWriter clones and does not retain the provided span set.
```
Something happened with the comment.

iskettaneh committed:
kvserver: add spanSetWriteBatch to spanset package

This commit adds spanSetWriteBatch to the spanset package and
asserts on accesses. The next commit will integrate it into
the SpansetBatch struct.

iskettaneh reviewed (commented):
I don't think that there is a reimplementation of some function. I split this PR into 2 commits!

> @iskettaneh made 3 comments and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159847#-:-Oh7tY5m9rD3JejHYJKt:bjkcnjg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv).



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -773,6 +774,108 @@ func NewReadWriterAt(rw storage.ReadWriter, spans *SpanSet, ts hlc.Timestamp) st
+	}
+	endKey, ok := storage.DecodeEngineKey(end)
+	if !ok {
+		return errors.Errorf("cannot decode start engine key")
```
`end`

same below

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -773,6 +774,108 @@ func NewReadWriterAt(rw storage.ReadWriter, spans *SpanSet, ts hlc.Timestamp) st
+	// Decode the internal key to check spans.
+	mvccKey, ok := storage.DecodeEngineKey(key.UserKey)
+	if !ok {
+		return errors.Errorf("cannot decode start engine key")
```
rm `start `

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -902,58 +916,14 @@ func (s spanSetBatch) NewBatchOnlyMVCCIterator(
 // spans.
 func (s spanSetBatch) shallowCopy() *spanSetBatch {
 	b := s
 	b.spanSetReader.spans = b.spanSetReader.spans.ShallowCopy()
```
A bit bummer that we need to do this twice, easy to forget. Not for this PR, but maybe something can be done.

iskettaneh replied:
I realized something while woking on this.
Initially, all spans are basically the same. However, when we do this shallow copy, we will have 3 different pointers. I now changed the logic to keep the 3 spans in sync. This makes the DisableSpanset checks a bit simpler.
<!-- Reviewable comment -OhAdYTHBfWm3kAWsDrZ:berbiob -->
<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
Looks good / mergeable, thanks for the commit split. Dumping easy nits, and will review the bottom half of the second commit in the morning with fresher head.

pav-kv commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -962,23 +932,33 @@ func (s spanSetBatch) shallowCopy() *spanSetBatch {
-		b:          b,
-		spans:      spans,
-		spansOnly:  true,
+		spanSetReader: spanSetReader{r: b, spans: spans, spansOnly: true},
```
We could then also avoid duplication like in `NewBatch / NewBatchAt` with:

```golang
func NewBatch(b storage.Batch, spans *SpanSet) {
	return newBatch(b, makeSpansStuff(spans, true, hlc.Timestamp{}))
}

func NewBatchAt(b storage.Batch, spans *SpanSet, ts hlc.Timestamp) {
	return newBatch(b, makeSpansStuff(spans, false, ts))
}

func newBatch(b, stuff) *spanSetBatch {
	return &spanSetBatch {
		// big repetitive stanza
	}
}
```

pav-kv replied:
We could probably also minimize the "shallow copy" stuff in some way:

```golang
func (s spanStuff) shallowCopy() spanStuff { }

func (s spanSetBatch) shallowCopy() spanSetBatch {
	// NB: all the "make" variants don't allocate
	return makeSpanSetBatch(s.b, s.spansStuff.shallowCopy())
}
```

iskettaneh replied:
I have a plan to do it in a future PR. I agree that I don't like that we are keeping track of multiple spans, that are all the same basically
<!-- Reviewable comment -OhAeE0UETXCruayL6Uu:bi33k2i -->
<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
LGTM. Left one suggestion at the bottom on how to reduce repetitions. If something like this makes sense in a broader context, can do it in a separate PR for all types rather than one.

iskettaneh commented on pkg/kv/kvserver/spanset/batch.go:
```diff
@@ -902,58 +916,14 @@ func (s spanSetBatch) NewBatchOnlyMVCCIterator(
 	b := s
 	b.spanSetReader.spans = b.spanSetReader.spans.ShallowCopy()
-	b.spanSetWriter.spans = b.spanSetWriter.spans.ShallowCopy()
+	b.spanSetWriteBatch.spanSetWriter.spans = b.spanSetWriteBatch.spanSetWriter.spans.ShallowCopy()
```
No, I underestimated Go. I now reverted it back
<!-- Reviewable comment -OhAdpaUBZYfrZgFV37n:b-47sfrv -->
<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> @iskettaneh made 7 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159847)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv).



<!-- Sent from Reviewable.io -->
